File: ch7/7-3.c
#include <stdio.h>
#include <stdarg.h>

void miniprintf(char * fmt, ...);

int main(void)
{
    miniprintf("This is the %c\n", 's');
}

void miniprintf(char * fmt, ...)
{
    va_list ap;
    char * p, * sval;
    int ival;
    double dval;

    va_start(ap, fmt);
    for (p = fmt; *p; p++) {
        if (*p != '%') {
            putchar( * p);
            continue;
        }
        switch (*++p) {
            case 'i':
            case 'd':
                ival = va_arg(ap, int);
                printf("%d", ival);
                break;
            case 'o':
                ival = va_arg(ap, int);
                printf("%o", ival);
                break;
            case 'X':
            case 'x':
                ival = va_arg(ap, int);
                printf("%x", ival);
                break;
            case 'u':
                ival = va_arg(ap, int);
                printf("%u", ival);
                break;
            case 'c':
                ival = va_arg(ap, int);
                printf("%c", ival);
                break;
            case 'G':
            case 'g':
            case 'E':
            case 'e':
            case 'f':
                dval = va_arg(ap, double);
                printf("%f", dval);
                break;
            case 's':
                for (sval = va_arg(ap, char *); *sval; sval++) {
                    putchar(*sval);
                }
                break;
            default:
                putchar(*p);
                break;
        }
    }
    va_end(ap);
}

File: ch7/7-3.go
package main

import "fmt"

func main() {
	miniprintf("This is the letter %c\n", 's')
}

func miniprintf(fm string, ap ...interface{}) {
	var j int
	for i, p := range fm {
		if i > 0 && fm[i-1] == '%' {
			continue
		}
		if p != '%' {
			fmt.Printf("%c", p)
			continue
		}
		v := ap[j]
		switch fm[i+1] {
		case 'i', 'd':
			fmt.Printf("%d", v.(int))
			j++
			break
		case 'o':
			fmt.Printf("%o", v.(int))
			j++
			break
		case 'X', 'x':
			fmt.Printf("%x", v.(int))
			j++
			break
		case 'u':
			fmt.Printf("%u", v.(int))
			j++
			break
		case 'c':
			fmt.Printf("%c", v.(rune))
			j++
			break
		case 'G', 'g', 'E', 'e', 'f':
			fmt.Printf("%f", v.(float64))
			j++
			break
		case 's':
			fmt.Printf("%s", v.(string))
			j++
			break
		default:
			fmt.Printf("%v", v)
			j++
			break

		}
	}
}
